In [1]:
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://postgres:N9xmzrzdhs*@localhost:5432/carsales"
)


In [2]:
raw = pd.read_csv("data/Car Sales.xlsx - car_data.csv")
raw.head()

car_id date_of_sale customer_name gender  annual_income  \
0  C_CND_000001       1/2/22     Geraldine   Male          13500   
1  C_CND_000002       1/2/22           Gia   Male        1480000   
2  C_CND_000003       1/2/22        Gianna   Male        1035000   
3  C_CND_000004       1/2/22       Giselle   Male          13500   
4  C_CND_000005       1/2/22         Grace   Male        1465000   

                           dealer_name   company       model  \
0  Buddy Storbeck's Diesel Service Inc      Ford  Expedition   
1                     C & M Motors Inc     Dodge     Durango   
2                          Capitol KIA  Cadillac    Eldorado   
3               Chrysler of Tri-Cities    Toyota      Celica   
4                    Chrysler Plymouth     Acura          TL   

                      engine transmission      colour  price   dealer_no  \
0  Double Overhead Camshaft         Auto       Black  26000  06457-3834   
1  Double Overhead Camshaft         Auto       Black  19000  60504-7114   
2          Overhead Camshaft       Manual         Red  31500  38701-8047   
3          Overhead Camshaft       Manual  Pale White  14000  99301-3882   
4  Double Overhead Camshaft         Auto         Red  24500  53546-9427   

  body_style    phone dealer_region  
0        SUV  8264678    Middletown  
1        SUV  6848189        Aurora  
2  Passenger  7298798    Greenville  
3        SUV  6257557         Pasco  
4  Hatchback  7081483    Janesville

In [3]:
raw.to_sql("car_sales", engine, if_exists="replace", index=False)

906

In [4]:
def read_sql(query: str) -> pd.DataFrame:
    df = pd.read_sql_query(query, engine)
    df.head()
    return df

## Data quality check

No missing values (see below).

In [5]:
read_sql(
"""
SELECT
    COUNT(*) AS total_rows,
    SUM(CASE WHEN car_id IS NULL          THEN 1 ELSE 0 END) AS null_car_id,
    SUM(CASE WHEN date_of_sale IS NULL    THEN 1 ELSE 0 END) AS null_date,
    SUM(CASE WHEN customer_name IS NULL   THEN 1 ELSE 0 END) AS null_customer_name,
    SUM(CASE WHEN annual_income IS NULL   THEN 1 ELSE 0 END) AS null_annual_income,
    SUM(CASE WHEN dealer_name IS NULL     THEN 1 ELSE 0 END) AS null_dealer_name,
    SUM(CASE WHEN dealer_region IS NULL   THEN 1 ELSE 0 END) AS null_dealer_region,
    SUM(CASE WHEN price IS NULL           THEN 1 ELSE 0 END) AS null_price
FROM car_sales;
"""
)


total_rows  null_car_id  null_date  null_customer_name  null_annual_income  \
0       23906            0          0                   0                   0   

   null_dealer_name  null_dealer_region  null_price  
0                 0                   0           0

No duplicates (see below).

In [6]:
read_sql("""
SELECT
    car_id,
    date_of_sale,
    customer_name,
    gender,
    annual_income,
    dealer_name,
    dealer_no,
    company,
    model,
    engine,
    transmission,
    colour,
    price,
    body_style,
    phone,
    dealer_region,
    COUNT(*) AS row_count
FROM car_sales
GROUP BY
    car_id, date_of_sale, customer_name, gender, annual_income,
    dealer_name, dealer_no, company, model, engine, transmission,
    colour, price, body_style, phone, dealer_region
HAVING COUNT(*) > 1;
""")

Empty DataFrame
Columns: [car_id, date_of_sale, customer_name, gender, annual_income, dealer_name, dealer_no, company, model, engine, transmission, colour, price, body_style, phone, dealer_region, row_count]
Index: []

In [7]:
read_sql("""
SELECT
    dealer_no,
    COUNT(DISTINCT dealer_name)  AS distinct_names,
    COUNT(DISTINCT dealer_region) AS distinct_regions
FROM car_sales
GROUP BY dealer_no
HAVING COUNT(DISTINCT dealer_name) > 1
    OR COUNT(DISTINCT dealer_region) > 1;
""")

dealer_no  distinct_names  distinct_regions
0  06457-3834               4                 7
1  38701-8047               4                 7
2  53546-9427               4                 7
3  60504-7114               4                 7
4  78758-7841               4                 7
5  85257-3102               4                 7
6  99301-3882               4                 7

In [8]:
read_sql("""
SELECT DISTINCT
	dealer_no,
	dealer_name,
	dealer_region
FROM car_sales
WHERE dealer_no IN (
    SELECT
        dealer_no
    FROM car_sales
    GROUP BY dealer_no
    HAVING COUNT(DISTINCT dealer_name) > 1
        OR COUNT(DISTINCT dealer_region) > 1
)
ORDER BY dealer_no, dealer_name, dealer_region;
""")

dealer_no                          dealer_name dealer_region
0    06457-3834  Buddy Storbeck's Diesel Service Inc        Aurora
1    06457-3834  Buddy Storbeck's Diesel Service Inc        Austin
2    06457-3834  Buddy Storbeck's Diesel Service Inc    Greenville
3    06457-3834  Buddy Storbeck's Diesel Service Inc    Janesville
4    06457-3834  Buddy Storbeck's Diesel Service Inc    Middletown
..          ...                                  ...           ...
191  99301-3882                 Star Enterprises Inc    Greenville
192  99301-3882                 Star Enterprises Inc    Janesville
193  99301-3882                 Star Enterprises Inc    Middletown
194  99301-3882                 Star Enterprises Inc         Pasco
195  99301-3882                 Star Enterprises Inc    Scottsdale

[196 rows x 3 columns]

In [9]:
read_sql("""
SELECT COUNT(DISTINCT dealer_no)
FROM car_sales;
""")

count
0      7

There are only 7 dealer numbesrs, so ithere should be no error, but we cannot use dealer_no to uniquely identify each dealer. Using dealer _name and dealer_region would be better.


First and last dates found in the dataset shown below.

In [10]:
read_sql("""
SELECT MIN(date_of_sale) AS first_sale, MAX(date_of_sale) as last_sale
FROM car_sales;
""")

first_sale last_sale
0    1/10/22    9/9/23

## Descriptive statistics

In [11]:
read_sql("""
SELECT
	company AS "Company",
	COUNT(DISTINCT model) AS "Number of models"
FROM car_sales
GROUP BY company
ORDER BY COUNT(DISTINCT model) DESC;
""")

Company  Number of models
0        Dodge                11
1         Ford                11
2       Toyota                 9
3    Chevrolet                 9
4   Mercedes-B                 9
5       Nissan                 7
6     Chrysler                 7
7   Mitsubishi                 7
8      Mercury                 6
9        Lexus                 6
10  Oldsmobile                 6
11       Volvo                 6
12  Volkswagen                 6
13      Saturn                 5
14       Honda                 5
15    Cadillac                 5
16     Pontiac                 5
17       Buick                 4
18    Plymouth                 4
19     Lincoln                 3
20        Audi                 3
21         BMW                 3
22     Hyundai                 3
23        Jeep                 3
24       Acura                 3
25     Porsche                 3
26      Subaru                 2
27        Saab                 2
28      Jaguar                 1
29    Infiniti                 1

In [12]:
read_sql("""
SELECT
    AVG(price)  AS avg_price,
    MIN(price)  AS min_price,
    MAX(price)  AS max_price,
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY price) AS p25_price,
    PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY price) AS median_price,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY price) AS p75_price
FROM car_sales;
""")

avg_price  min_price  max_price  p25_price  median_price  p75_price
0  28090.247846       1200      85800    18001.0       23000.0    34000.0

In [13]:
read_sql("""
SELECT
    AVG(annual_income) AS avg_income,
    MIN(annual_income) AS min_income,
    MAX(annual_income) AS max_income,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY annual_income) AS median_income
FROM car_sales;
""")

avg_income  min_income  max_income  median_income
0  830840.285117       10080    11200000       735000.0

In [14]:
read_sql("""
SELECT
    body_style,
    COUNT(*) AS units,
    COUNT(*)::numeric / SUM(COUNT(*)) OVER () AS share_of_total
FROM car_sales
GROUP BY body_style
ORDER BY units DESC;
""")

body_style  units  share_of_total
0        SUV   6374        0.266628
1  Hatchback   6128        0.256337
2      Sedan   4488        0.187735
3  Passenger   3945        0.165021
4    Hardtop   2971        0.124278

In [15]:
read_sql("""
SELECT
    engine,
    transmission,
    COUNT(*) AS units,
    COUNT(*)::numeric / SUM(COUNT(*)) OVER () AS share_of_total
FROM car_sales
GROUP BY engine, transmission
ORDER BY units DESC;
""")

engine transmission  units  share_of_total
0  Double Overhead Camshaft         Auto  12571        0.525851
1          Overhead Camshaft       Manual  11335        0.474149

## 16 Business Questions

### 1. What are the best and worst performing companies in terms of revenue and volume?

In [16]:
read_sql("""
SELECT 
	company AS "Company", 
	SUM(price) AS "Total revenue", 
	COUNT(price) AS "Units sold"
FROM car_sales
GROUP BY company
ORDER BY SUM(price) DESC
LIMIT 5;
""")

Company  Total revenue  Units sold
0   Chevrolet     47655265.0        1819
1        Ford     47231583.0        1614
2       Dodge     44124996.0        1671
3  Oldsmobile     35434512.0        1111
4  Mercedes-B     34624123.0        1285

Top 3 revenue are are Chevrolet, Ford, Dodge and Oldsmobile. American car companies are doing well.

In [17]:
read_sql("""
SELECT 
	company AS "Company", 
	SUM(price) AS "Total revenue", 
	COUNT(price) AS "Units sold"
FROM car_sales
GROUP BY company
ORDER BY COUNT(*) DESC
LIMIT 5;
""")

Company  Total revenue  Units sold
0   Chevrolet     47655265.0        1819
1       Dodge     44124996.0        1671
2        Ford     47231583.0        1614
3  Volkswagen     34082881.0        1333
4  Mercedes-B     34624123.0        1285

Top 3 revenue are are Chevrolet, Dodge and Ford. American car companies are doing well.\
Mercedes Benz is also 5th on both measures. Surprisingly, as I expected Japanese companies to have higher volume.

In [18]:
read_sql("""
SELECT 
	company AS "Company", 
	SUM(price) AS "Total revenue", 
	COUNT(price) AS "Units sold"
FROM car_sales
GROUP BY company
ORDER BY SUM(price)
LIMIT 5;
""")

Company  Total revenue  Units sold
0    Jaguar      4524875.0         180
1   Hyundai      5117966.0         264
2  Infiniti      5717040.0         195
3      Jeep      7643814.0         363
4      Saab      7668431.0         210

In [19]:
read_sql("""
SELECT 
	company AS "Company", 
	SUM(price) AS "Total revenue", 
	COUNT(price) AS "Units sold"
FROM car_sales
GROUP BY company
ORDER BY COUNT(*)
LIMIT 10;
""")

Company  Total revenue  Units sold
0    Jaguar      4524875.0         180
1  Infiniti      5717040.0         195
2      Saab      7668431.0         210
3   Hyundai      5117966.0         264
4   Porsche      8185637.0         361
5      Jeep      7643814.0         363
6    Subaru     11312193.0         405
7     Buick     14765485.0         439
8      Audi     10694638.0         468
9   Lincoln     15452262.0         492

Jaguar, Inifiniti, Saab, Hyundai, Porsche. Not that surprising since Porsche mostly sells sports cars which will not have high volumes. Actually, it's more surprising that there aren't more luxury brands (unless Jaguar is counted).\
\
Overall, high degree of correlation between companies with high (low) revenue and high (low) units sold, meaning that low revenues could be driven by low volume.\
Companies with more models in the dataset also tended to perform better than companies with fewer models.\
This makes sense since sufficient units need to be sold to jsutify the availability of a new model.

### 2. What are the best performing models in terms of revenue and volume?

In [21]:
read_sql("""
SELECT 
	company AS "Company", 
	model as "Model",
	SUM(price) AS "Total revenue", 
	COUNT(price) AS "Units sold"
FROM car_sales
GROUP BY company, model
ORDER BY SUM(price) DESC
LIMIT 5;
""")

Company          Model  Total revenue  Units sold
0       Lexus          LS400     14263424.0         354
1  Volkswagen          Jetta     12287736.0         382
2  Oldsmobile     Silhouette     12271849.0         411
3  Mitsubishi  Montero Sport     11472231.0         302
4       Dodge     Ram Pickup     10655391.0         383

The Lexus L400, Volkswagen Jetta, Oldsmobile Sillhouette, Mitsubishi Motero Sport and Dodge Ram Pickup generated the highest revnue. All of them had > 350 units sold.

In [22]:
read_sql("""
SELECT 
	company AS "Company", 
	model as "Model",
	SUM(price) AS "Total revenue", 
	COUNT(price) AS "Units sold"
FROM car_sales
GROUP BY company, model
ORDER BY COUNT(*) DESC
LIMIT 5;
""")

Company       Model  Total revenue  Units sold
0  Mitsubishi    Diamante      9308151.0         418
1  Oldsmobile  Silhouette     12271849.0         411
2   Chevrolet       Prizm      9386606.0         411
3  Volkswagen      Passat      7663885.0         391
4       Dodge  Ram Pickup     10655391.0         383

The Lexus Mitsubushi Diamante, Oldsmobile Sillhouette, Chevrolet Prizm, Volkswagen Passat and Dodge Ram Pickup had the most units sold.\
There is a high level of correlation between the best performing models in terms of revenue and volume.

### 3. What is the average selling price by company and body style, and which combinations are premium vs budget?

In [23]:
read_sql("""
WITH style_ranks AS (
	SELECT
		company,
		body_style,
		avg_price,
		price_rank_in_company / MAX(price_rank_in_company) OVER(
			PARTITION BY company
		) AS price_percentile_rank
	FROM (
		SELECT
			company,
			body_style,
			AVG(price) AS avg_price,
			RANK() OVER (
				PARTITION BY company
				ORDER BY AVG(price) DESC
			) AS price_rank_in_company
		FROM car_sales
		GROUP BY company, body_style
	) AS temp
)
SELECT
    body_style AS "Body style",
    AVG(price_percentile_rank) AS "Average rank across companies"
FROM style_ranks
GROUP BY body_style
ORDER BY AVG(price_percentile_rank);
""")

Body style  Average rank across companies
0    Hardtop                       0.130435
1      Sedan                       0.217391
2        SUV                       0.222222
3  Passenger                       0.291667
4  Hatchback                       0.310345

In [24]:
read_sql("""
SELECT company, model, body_style
FROM car_sales
WHERE body_style = 'Passenger';
""")

company        model body_style
0       Cadillac     Eldorado  Passenger
1         Toyota      Corolla  Passenger
2     Mitsubishi       Galant  Passenger
3           Ford       Escort  Passenger
4       Cadillac     Eldorado  Passenger
...          ...          ...        ...
3940      Jaguar       S-Type  Passenger
3941       Honda      Odyssey  Passenger
3942     Lincoln  Continental  Passenger
3943    Plymouth      Voyager  Passenger
3944   Chevrolet        Metro  Passenger

[3945 rows x 3 columns]

In [25]:
read_sql("""
SELECT company, model, body_style
FROM car_sales
WHERE body_style = 'Hardtop';
""")

company          model body_style
0      Chevrolet         Malibu    Hardtop
1         Nissan     Pathfinder    Hardtop
2       Infiniti            I30    Hardtop
3           Audi             A4    Hardtop
4           Audi             A4    Hardtop
...          ...            ...        ...
2966       Dodge        Caravan    Hardtop
2967       Dodge     Ram Pickup    Hardtop
2968  Mitsubishi  Montero Sport    Hardtop
2969   Chevrolet          Prizm    Hardtop
2970       Lexus          ES300    Hardtop

[2971 rows x 3 columns]

In [26]:
read_sql("""
SELECT company, model, body_style
FROM car_sales
WHERE body_style = 'SUV';
""")

company          model body_style
0          Ford     Expedition        SUV
1         Dodge        Durango        SUV
2        Toyota         Celica        SUV
3         Acura             RL        SUV
4       Mercury  Grand Marquis        SUV
...         ...            ...        ...
6369      Acura             RL        SUV
6370      Honda           CR-V        SUV
6371  Chevrolet       Corvette        SUV
6372        BMW           528i        SUV
6373      Dodge          Viper        SUV

[6374 rows x 3 columns]

Overall, not super indicative of which body styles should be counted as more popular overall, but useful in understanding the tastes in the regions in the dataset.

### 4. Which companies are premium and which are budget?

In [27]:
read_sql("""
SELECT
	company AS "Company",
	AVG(price) AS "Higher average price"
FROM car_sales
GROUP BY company
ORDER BY AVG(price) DESC;
""")

Company  Higher average price
0     Cadillac          40972.093558
1         Saab          36516.338095
2        Lexus          34024.567332
3        Buick          33634.362187
4   Oldsmobile          31894.250225
5      Lincoln          31407.036585
6       Saturn          31092.609215
7       Toyota          29513.120721
8     Plymouth          29404.980551
9      Pontiac          29358.300251
10    Infiniti          29318.153846
11        Ford          29263.682156
12     Mercury          28535.163616
13       Honda          28082.959040
14      Subaru          27931.340741
15       Volvo          27788.593156
16      Nissan          27047.511287
17  Mercedes-B          26944.842802
18  Mitsubishi          26673.818324
19       Dodge          26406.341113
20   Chevrolet          26198.606377
21    Chrysler          26019.529464
22  Volkswagen          25568.552888
23      Jaguar          25138.194444
24         BMW          25090.622785
25       Acura          24758.561684
26        Audi          22851.790598
27     Porsche          22674.894737
28        Jeep          21057.338843
29     Hyundai          19386.234848

It's not surprising that Cadillac, Saab, Buick and Lexus are on top. But I'm surprised that Porsche and Jeep rank so low, and all of these German brands are ranking lower than Toyota and Infiniti which are more budget options.

### 5. How about the most expensive models?

In [28]:
read_sql("""
SELECT
	company AS "Company",
	model AS "Model",
	AVG(price) AS "Average price"
FROM car_sales
GROUP BY company, model
ORDER BY AVG(price) DESC;
""")

Company    Model  Average price
0      Cadillac   Catera   56836.933333
1      Cadillac  DeVille   46154.050000
2          Ford  Contour   45309.079646
3         Acura  Integra   44604.000000
4        Toyota   Tacoma   44576.743017
..          ...      ...            ...
150  Mitsubishi  Montero   16035.000000
151  Mercedes-B  C-Class   14538.414474
152        Ford   Escort   12901.092105
153    Chrysler   Cirrus   12727.243243
154  Mitsubishi   Mirage    9063.368421

[155 rows x 3 columns]

The models and the body styles are the main factors driving the differnece in prices. For example, Toyota's most expensive model is the Tacoma which is a pickup truck and definitely more expensive than say the Audi A4.\
Again, this reflects the consumption habits of the drivers in these regions (some cities in the US).

### 6. Which cars are stauts symbols, i.e. driven by the richest?

In [29]:
read_sql("""
SELECT
	company AS "Company",
	model AS "Model",
	AVG(annual_income) AS "Annual income",
	AVG(price) AS "Average price"
FROM car_sales
GROUP BY company, model
ORDER BY AVG(annual_income) DESC;
""")

Company           Model  Annual income  Average price
0      Chrysler   Sebring Conv.   1.196270e+06   16250.100000
1       Hyundai          Accent   1.017677e+06   18129.085714
2        Toyota          Celica   1.010119e+06   25755.325843
3          Saab           5-Sep   1.006603e+06   30070.769231
4       Porsche  Carrera Cabrio   9.945521e+05   26577.312500
..          ...             ...            ...            ...
150       Lexus           RX300   6.718667e+05   18020.133333
151    Chrysler          Cirrus   6.663378e+05   12727.243243
152      Toyota          Avalon   6.652000e+05   18000.133333
153  Mitsubishi          Mirage   6.365000e+05    9063.368421
154       Acura         Integra   6.270688e+05   44604.000000

[155 rows x 4 columns]

Not the most expensive models. Usually not even the most expensive models from each company.

In [30]:
read_sql("""
SELECT
	company AS "Company",
	AVG(annual_income) AS "Annual income",
	AVG(price) AS "Average price"
FROM car_sales
GROUP BY company
ORDER BY AVG(annual_income) DESC;
""")

Company  Annual income  Average price
0     Cadillac  913476.457055   40972.093558
1      Hyundai  894106.772727   19386.234848
2         Saab  892487.771429   36516.338095
3      Mercury  890657.229977   28535.163616
4       Toyota  876490.990090   29513.120721
5       Nissan  863774.747178   27047.511287
6     Infiniti  861800.923077   29318.153846
7        Buick  855929.840547   33634.362187
8   Volkswagen  854321.878470   25568.552888
9      Porsche  853423.822715   22674.894737
10       Volvo  837570.499366   27788.593156
11    Plymouth  834670.745543   29404.980551
12       Dodge  834585.241771   26406.341113
13  Mitsubishi  833272.748630   26673.818324
14    Chrysler  833011.186607   26019.529464
15         BMW  831186.075949   25090.622785
16  Oldsmobile  830651.319532   31894.250225
17  Mercedes-B  823098.915953   26944.842802
18        Audi  819655.626068   22851.790598
19      Saturn  816950.767918   31092.609215
20        Jeep  814149.966942   21057.338843
21        Ford  813954.126394   29263.682156
22       Acura  810430.190131   24758.561684
23   Chevrolet  799263.693238   26198.606377
24      Subaru  795085.550617   27931.340741
25     Lincoln  794274.695122   31407.036585
26     Pontiac  790448.979899   29358.300251
27       Honda  778164.860169   28082.959040
28       Lexus  776040.334165   34024.567332
29      Jaguar  732528.544444   25138.194444

In terms of companies, the highest annual income seems highly related to the companies with the highest price.\
This suggests that the pirce may be driven by the tastetes of peopel with higher purchasing power or that higher average prices turn the company into a status symbol sought after by those who can afford it.


### 7. Instead of analysising in this way, let's take the top 10% and 20% in income and find the most popular cars among them.

In [31]:
read_sql("""
WITH income_percentiles AS (
    SELECT
        *,
        NTILE(10) OVER (ORDER BY annual_income) AS income_decile
    FROM car_sales
)
SELECT
    company AS "Company",
    model AS "Model",
    COUNT(*) AS "Units sold"
FROM income_percentiles
WHERE income_decile = 10
GROUP BY company, model
ORDER BY COUNT(*) DESC;
""")

Company       Model  Units sold
0    Volkswagen      Passat          46
1    Mitsubishi    Diamante          43
2    Oldsmobile  Silhouette          43
3         Dodge  Ram Pickup          42
4    Volkswagen       Jetta          42
..          ...         ...         ...
149       Volvo         V40           2
150       Lexus       RX300           1
151      Toyota      Avalon           1
152       Acura     Integra           1
153  Oldsmobile       Alero           1

[154 rows x 3 columns]

Though these models did not appear very high in the average price list, they were seen very high in the total revnue list, suggesting that higher quantitiates of these purcharsed led to higher revenue, as opposed to price.

In [32]:
read_sql("""
WITH income_percentiles AS (
    SELECT
        *,
        NTILE(10) OVER (ORDER BY annual_income) AS income_decile
    FROM car_sales
)
SELECT
    company AS "Company",
    model AS "Model",
    COUNT(*) AS "Units sold"
FROM income_percentiles
WHERE income_decile >= 9
GROUP BY company, model
ORDER BY COUNT(*) DESC;
""")

Company       Model  Units sold
0    Mitsubishi    Diamante          96
1    Oldsmobile  Silhouette          89
2         Dodge  Ram Pickup          81
3    Volkswagen       Jetta          81
4    Volkswagen      Passat          79
..          ...         ...         ...
150  Mitsubishi      Mirage           3
151       Dodge        Neon           3
152       Lexus       RX300           2
153      Toyota      Avalon           2
154  Oldsmobile       Alero           2

[155 rows x 3 columns]

### 8. Which dealer_regions attract the highest‑income customers on average?

In [33]:
read_sql("""
SELECT
	dealer_region AS "Dealer Region",
	AVG(annual_income) AS "Avergae income",
	COUNT(DISTINCT dealer_no) AS "Number of dealers"
FROM car_sales
GROUP BY dealer_region
ORDER BY AVG(annual_income) DESC;
""")

Dealer Region  Avergae income  Number of dealers
0         Pasco   853975.290323                  7
1        Aurora   845510.435783                  7
2    Janesville   833916.268778                  7
3    Greenville   832667.512788                  7
4    Middletown   825592.651854                  7
5        Austin   822849.572189                  7
6    Scottsdale   805682.702884                  7

The difference in average income between the countries is quite small. Same numebr of dealers in each region (7).

### 9. Best-performing dealers?

In [34]:
read_sql("""
SELECT 
	dealer_name AS "Dealer name",
    dealer_no AS "Dealer number", 
    dealer_region AS "Dealer region", 
	COUNT(*) AS "Units sold"
FROM car_sales
GROUP BY dealer_region, dealer_name, dealer_no
ORDER BY COUNT(*) DESC;
""")

Dealer name Dealer number  \
0    Progressive Shippers Cooperative Association No    53546-9427   
1                                      Race Car Help    78758-7841   
2                               Star Enterprises Inc    99301-3882   
3                                   Saab-Belle Dodge    60504-7114   
4                                          U-Haul CO    78758-7841   
..                                               ...           ...   
191                      Motor Vehicle Branch Office    78758-7841   
192                     McKinney Dodge Chrysler Jeep    85257-3102   
193                                  Iceberg Rentals    53546-9427   
194                              Hatfield Volkswagen    99301-3882   
195                                    Classic Chevy    85257-3102   

    Dealer region  Units sold  
0      Janesville        1069  
1          Austin        1019  
2           Pasco        1009  
3          Aurora        1009  
4          Austin        1008  
..            ...         ...  
191         Pasco          11  
192    Greenville          10  
193        Aurora          10  
194    Middletown          10  
195    Middletown           9  

[196 rows x 4 columns]

In [35]:
read_sql("""
SELECT 
	dealer_name AS "Dealer name",
    dealer_no AS "Dealer number", 
	dealer_region AS "Dealer region",
	SUM(price) AS "Total revenue"
FROM car_sales
GROUP BY dealer_region, dealer_name, dealer_no
ORDER BY SUM(price) DESC;
""")

Dealer name Dealer number  \
0    Progressive Shippers Cooperative Association No    53546-9427   
1                  Scrivener Performance Engineering    38701-8047   
2                                          U-Haul CO    78758-7841   
3                                   Saab-Belle Dodge    60504-7114   
4                     Ryder Truck Rental and Leasing    06457-3834   
..                                               ...           ...   
191                            Enterprise Rent A Car    60504-7114   
192                                  Iceberg Rentals    53546-9427   
193                                    Classic Chevy    85257-3102   
194                     McKinney Dodge Chrysler Jeep    85257-3102   
195                               Diehl Motor CO Inc    06457-3834   

    Dealer region  Total revenue  
0      Janesville     29795259.0  
1      Greenville     28880779.0  
2          Austin     28578406.0  
3          Aurora     28532977.0  
4      Middletown     28290409.0  
..            ...            ...  
191         Pasco       267651.0  
192        Aurora       258601.0  
193    Middletown       252300.0  
194    Greenville       230601.0  
195    Greenville       218501.0  

[196 rows x 4 columns]

Some overlap between the 2 measures. There's no outstanding region in this case.


### 10. Variety of models and conentration in each region?

In [36]:
read_sql("""
SELECT
    dealer_name AS "Dealer name",
    dealer_no AS "Dealer number",
    COUNT(DISTINCT model) AS "Distinct models sold",
    COUNT(*) AS "Units sold"
FROM car_sales
GROUP BY dealer_name, dealer_no
ORDER BY COUNT(DISTINCT model) DESC;
""")

Dealer name Dealer number  \
0                                Tri-State Mack Inc    85257-3102   
1                 Scrivener Performance Engineering    38701-8047   
2                              Star Enterprises Inc    99301-3882   
3   Progressive Shippers Cooperative Association No    53546-9427   
4                                     Race Car Help    78758-7841   
5                    Ryder Truck Rental and Leasing    06457-3834   
6                              Rabun Used Car Sales    85257-3102   
7                                         U-Haul CO    78758-7841   
8                       Motor Vehicle Branch Office    78758-7841   
9                                     Suburban Ford    53546-9427   
10                                 Saab-Belle Dodge    60504-7114   
11                                  Iceberg Rentals    53546-9427   
12                               Diehl Motor CO Inc    06457-3834   
13                                 C & M Motors Inc    60504-7114   
14                                   Nebo Chevrolet    06457-3834   
15                          Clay Johnson Auto Sales    78758-7841   
16                  New Castle Ford Lincoln Mercury    60504-7114   
17                                    Classic Chevy    85257-3102   
18                           Chrysler of Tri-Cities    99301-3882   
19              Buddy Storbeck's Diesel Service Inc    06457-3834   
20                       Gartner Buick Hyundai Saab    38701-8047   
21                                      Capitol KIA    38701-8047   
22                                  Pars Auto Sales    38701-8047   
23            Pitre Buick-Pontiac-Gmc of Scottsdale    99301-3882   
24                                Chrysler Plymouth    53546-9427   
25                              Hatfield Volkswagen    99301-3882   
26                     McKinney Dodge Chrysler Jeep    85257-3102   
27                            Enterprise Rent A Car    60504-7114   

    Distinct models sold  Units sold  
0                    152        1249  
1                    151        1246  
2                    150        1249  
3                    150        1318  
4                    149        1253  
5                    149        1248  
6                    149        1313  
7                    148        1247  
8                    147         626  
9                    147        1243  
10                   146        1251  
11                   146         627  
12                   146         624  
13                   146         625  
14                   145         633  
15                   145         627  
16                   145         629  
17                   144         623  
18                   144         626  
19                   143         627  
20                   143         628  
21                   143         628  
22                   143         630  
23                   143         628  
24                   142         625  
25                   141         629  
26                   139         629  
27                   139         625

Quite a lot of verlap with the highest volume and highest revenue ones.\
Suggetst that the best strategy for having higher revenue is to diversifying products sold instead of focusing on certain models only.


In [37]:
read_sql("""
WITH dealer_model_counts AS (
    SELECT
        dealer_name,
        dealer_no,
        model,
        COUNT(*) AS model_sales
    FROM car_sales
    GROUP BY dealer_name, dealer_no, model
),
dealer_totals AS (
    SELECT
        dealer_name,
        dealer_no,
        SUM(model_sales) AS total_sales
    FROM dealer_model_counts
    GROUP BY dealer_name, dealer_no
),
dealer_model_share AS (
    SELECT
        dmc.dealer_name,
        dmc.dealer_no,
        dmc.model,
        dmc.model_sales,
        dt.total_sales,
        dmc.model_sales::decimal / dt.total_sales AS model_share,
        RANK() OVER (
            PARTITION BY dmc.dealer_name, dmc.dealer_no
            ORDER BY dmc.model_sales DESC
        ) AS model_rank_in_dealer
    FROM dealer_model_counts dmc
    JOIN dealer_totals dt
      ON dmc.dealer_name = dt.dealer_name
     AND dmc.dealer_no   = dt.dealer_no
)
SELECT
    dealer_name AS "Dealer name",
    dealer_no AS "Dealer number",
    COUNT(DISTINCT model) AS "Distincts models sold",
    SUM(CASE WHEN model_rank_in_dealer <= 3 THEN model_share ELSE 0 END)
        AS "Top 3 model share"
FROM dealer_model_share
GROUP BY dealer_name, dealer_no
ORDER BY "Top 3 model share" DESC;
""")

Dealer name Dealer number  \
0                                   Iceberg Rentals    53546-9427   
1                                       Capitol KIA    38701-8047   
2                                    Nebo Chevrolet    06457-3834   
3               Buddy Storbeck's Diesel Service Inc    06457-3834   
4                                     Suburban Ford    53546-9427   
5                           Clay Johnson Auto Sales    78758-7841   
6                    Ryder Truck Rental and Leasing    06457-3834   
7             Pitre Buick-Pontiac-Gmc of Scottsdale    99301-3882   
8                                         U-Haul CO    78758-7841   
9                                     Classic Chevy    85257-3102   
10  Progressive Shippers Cooperative Association No    53546-9427   
11                     McKinney Dodge Chrysler Jeep    85257-3102   
12                               Diehl Motor CO Inc    06457-3834   
13                      Motor Vehicle Branch Office    78758-7841   
14                Scrivener Performance Engineering    38701-8047   
15                             Star Enterprises Inc    99301-3882   
16                                    Race Car Help    78758-7841   
17                                  Pars Auto Sales    38701-8047   
18                             Rabun Used Car Sales    85257-3102   
19                  New Castle Ford Lincoln Mercury    60504-7114   
20                            Enterprise Rent A Car    60504-7114   
21                                 C & M Motors Inc    60504-7114   
22                                Chrysler Plymouth    53546-9427   
23                              Hatfield Volkswagen    99301-3882   
24                       Gartner Buick Hyundai Saab    38701-8047   
25                                 Saab-Belle Dodge    60504-7114   
26                               Tri-State Mack Inc    85257-3102   
27                           Chrysler of Tri-Cities    99301-3882   

    Distincts models sold  Top 3 model share  
0                     146           0.133971  
1                     143           0.128981  
2                     145           0.104265  
3                     143           0.098884  
4                     147           0.096541  
5                     145           0.095694  
6                     149           0.093750  
7                     143           0.090764  
8                     148           0.089014  
9                     144           0.085072  
10                    150           0.083460  
11                    139           0.081081  
12                    146           0.076923  
13                    147           0.076677  
14                    151           0.075441  
15                    150           0.075260  
16                    149           0.072626  
17                    143           0.066667  
18                    149           0.064737  
19                    145           0.063593  
20                    139           0.062400  
21                    146           0.062400  
22                    142           0.062400  
23                    141           0.062003  
24                    143           0.060510  
25                    146           0.058353  
26                    152           0.057646  
27                    144           0.049521

The dealers who have high volume and revenue are all quite average in terms of the concentration of the top 3 models sold. \
This suggestst that concentration does not really matter.

### 11. Which dealers have the greatest discounts by region? (Proportion of average price regionally)

In [38]:
read_sql("""
WITH regional_model_avg AS (
    SELECT
        dealer_region,
        model,
        AVG(price) AS regional_avg_price
    FROM car_sales
    GROUP BY dealer_region, model
),
sales_with_discount AS (
    SELECT
        c.dealer_name,
        c.dealer_no,
        c.dealer_region,
        c.model,
        c.price,
        rma.regional_avg_price,
        (c.price - rma.regional_avg_price) / rma.regional_avg_price AS price_diff_prop
    FROM car_sales c
    JOIN regional_model_avg rma
      ON c.dealer_region = rma.dealer_region
     AND c.model         = rma.model
)
SELECT
    dealer_name AS "Dealer name",
    dealer_no AS "Dealer number",
    AVG(price_diff_prop) AS "Average discount vs region"
FROM sales_with_discount
GROUP BY dealer_name, dealer_no
ORDER BY "Average discount vs region";
""")

Dealer name Dealer number  \
0                                    Nebo Chevrolet    06457-3834   
1               Buddy Storbeck's Diesel Service Inc    06457-3834   
2                           Clay Johnson Auto Sales    78758-7841   
3                                     Race Car Help    78758-7841   
4                                 Chrysler Plymouth    53546-9427   
5                                Tri-State Mack Inc    85257-3102   
6                      McKinney Dodge Chrysler Jeep    85257-3102   
7                                  Saab-Belle Dodge    60504-7114   
8                                       Capitol KIA    38701-8047   
9                 Scrivener Performance Engineering    38701-8047   
10            Pitre Buick-Pontiac-Gmc of Scottsdale    99301-3882   
11                      Motor Vehicle Branch Office    78758-7841   
12                                  Pars Auto Sales    38701-8047   
13                              Hatfield Volkswagen    99301-3882   
14  Progressive Shippers Cooperative Association No    53546-9427   
15                             Star Enterprises Inc    99301-3882   
16                   Ryder Truck Rental and Leasing    06457-3834   
17                                 C & M Motors Inc    60504-7114   
18                            Enterprise Rent A Car    60504-7114   
19                               Diehl Motor CO Inc    06457-3834   
20                                    Suburban Ford    53546-9427   
21                           Chrysler of Tri-Cities    99301-3882   
22                  New Castle Ford Lincoln Mercury    60504-7114   
23                                    Classic Chevy    85257-3102   
24                             Rabun Used Car Sales    85257-3102   
25                       Gartner Buick Hyundai Saab    38701-8047   
26                                        U-Haul CO    78758-7841   
27                                  Iceberg Rentals    53546-9427   

    Average discount vs region  
0                    -0.015581  
1                    -0.015543  
2                    -0.012902  
3                    -0.012607  
4                    -0.012458  
5                    -0.012174  
6                    -0.010181  
7                    -0.009642  
8                    -0.008398  
9                    -0.007299  
10                   -0.006578  
11                   -0.004895  
12                   -0.003690  
13                   -0.003332  
14                    0.000330  
15                    0.002669  
16                    0.002889  
17                    0.003014  
18                    0.003081  
19                    0.004626  
20                    0.008163  
21                    0.009720  
22                    0.010515  
23                    0.013051  
24                    0.013413  
25                    0.013804  
26                    0.020909  
27                    0.021592

Even though they offer great discounts, Race Car Help, Saab-Belle Dodge and Scrivener Performance Engineering have some of the highest revenue, meaning that these discounts are able to bring in the volume needed to overcome the lowered prices.


### 12. Revenue share from each income quartile for each dealer.

In [39]:
read_sql("""
WITH income_quantiles AS (
    SELECT
        dealer_name,
        dealer_no,
        price,
        NTILE(4) OVER (ORDER BY annual_income) AS income_quartile
    FROM car_sales
),
dealer_revenue AS (
    SELECT
        dealer_name,
        dealer_no,
        income_quartile,
        SUM(price) AS revenue_by_quartile
    FROM income_quantiles
    GROUP BY dealer_name, dealer_no, income_quartile
),
dealer_total AS (
    SELECT
        dealer_name,
        dealer_no,
        SUM(revenue_by_quartile) AS total_revenue
    FROM dealer_revenue
    GROUP BY dealer_name, dealer_no
)
SELECT
    dr.dealer_name AS "Dealer name",
    dr.dealer_no AS "Dealer number",
    dr.income_quartile AS "Income quartile",
    dr.revenue_by_quartile AS "Total revenue",
    dr.revenue_by_quartile::numeric / dt.total_revenue AS "Revenue share"
FROM dealer_revenue dr
JOIN dealer_total dt
  ON dr.dealer_name = dt.dealer_name
 AND dr.dealer_no   = dt.dealer_no
ORDER BY dr.dealer_name, dr.income_quartile;
""")

Dealer name Dealer number  Income quartile  \
0    Buddy Storbeck's Diesel Service Inc    06457-3834                1   
1    Buddy Storbeck's Diesel Service Inc    06457-3834                2   
2    Buddy Storbeck's Diesel Service Inc    06457-3834                3   
3    Buddy Storbeck's Diesel Service Inc    06457-3834                4   
4                       C & M Motors Inc    60504-7114                1   
..                                   ...           ...              ...   
107                   Tri-State Mack Inc    85257-3102                4   
108                            U-Haul CO    78758-7841                1   
109                            U-Haul CO    78758-7841                2   
110                            U-Haul CO    78758-7841                3   
111                            U-Haul CO    78758-7841                4   

     Total revenue  Revenue share  
0        4168531.0       0.244271  
1        3858235.0       0.226088  
2        4434915.0       0.259880  
3        4603542.0       0.269762  
4        4319041.0       0.245821  
..             ...            ...  
107      8523119.0       0.242884  
108      8943578.0       0.249291  
109      9042423.0       0.252046  
110      8917849.0       0.248574  
111      8972239.0       0.250090  

[112 rows x 5 columns]

Most revenue shares are between 20% and 30%.\
Hatfield Volkswagen has 30% of revenue share from 2nd quartile (and 26% from 4th quartile). This could be a coincidence as there is no obvious neglect of other income groups.\
Nebo Chevrolet has 19% of revenue share from 4th quartile. Its business is actually clearly concentrated in the lower income quartiles (28% in 1st, 30% in 2nd).\
We note that these dealers do not appear in the list of dealers with highest volume or revenue. This suggestst that diversifying the target audience of sales is a btter strategy than focusing on certain types of customers, just like is the case for car models. Furthermore, these 2 strategies are liekly to be highly correlated.


### 13. Growth rate of dealers from the first 6 months to the last 6 months.

In [ ]:
read_sql("""
WITH dealer_monthly AS (
    SELECT
        dealer_name,
        dealer_no,
        DATE_TRUNC('month', date_of_sale::date) AS month,
        SUM(price) AS revenue
    FROM car_sales
    WHERE date_of_sale::date BETWEEN DATE '2022-10-01' AND DATE '2023-09-30'
    GROUP BY dealer_name, dealer_no, DATE_TRUNC('month', date_of_sale::date)
),
dealer_halfyear AS (
    SELECT
        dealer_name,
        dealer_no,
        CASE
            WHEN month BETWEEN DATE '2022-10-01' AND DATE '2023-03-31'
                THEN 'H1'
            ELSE 'H2'
        END AS period,
        SUM(revenue) AS revenue
    FROM dealer_monthly
    GROUP BY dealer_name, dealer_no,
             CASE
                 WHEN month BETWEEN DATE '2022-10-01' AND DATE '2023-03-31'
                     THEN 'H1'
                 ELSE 'H2'
             END
),
dealer_pivot AS (
    SELECT
        dealer_name,
        dealer_no,
        SUM(CASE WHEN period = 'H1' THEN revenue ELSE 0 END) AS rev_h1,
        SUM(CASE WHEN period = 'H2' THEN revenue ELSE 0 END) AS rev_h2
    FROM dealer_halfyear
    GROUP BY dealer_name, dealer_no
)
SELECT
    dealer_name  AS "Dealer name",
    dealer_no    AS "Dealer number",
    rev_h1       AS "Revenue Oct 22–Mar 23",
    rev_h2       AS "Revenue Apr 23–Sep 23",
    rev_h2 - rev_h1 AS "Absolute growth",
    CASE
        WHEN rev_h1 = 0 THEN NULL
        ELSE (rev_h2 - rev_h1)::numeric / rev_h1
    END AS "Percentage growth"
FROM dealer_pivot
ORDER BY "Percentage growth" DESC NULLS LAST;

""")

Dealer name Dealer number  \
0                           Clay Johnson Auto Sales    78758-7841   
1                                   Iceberg Rentals    53546-9427   
2                              Rabun Used Car Sales    85257-3102   
3                                         U-Haul CO    78758-7841   
4             Pitre Buick-Pontiac-Gmc of Scottsdale    99301-3882   
5                                   Pars Auto Sales    38701-8047   
6                                Tri-State Mack Inc    85257-3102   
7                        Gartner Buick Hyundai Saab    38701-8047   
8                                 Chrysler Plymouth    53546-9427   
9                                  Saab-Belle Dodge    60504-7114   
10                              Hatfield Volkswagen    99301-3882   
11                      Motor Vehicle Branch Office    78758-7841   
12                  New Castle Ford Lincoln Mercury    60504-7114   
13                               Diehl Motor CO Inc    06457-3834   
14                     McKinney Dodge Chrysler Jeep    85257-3102   
15                Scrivener Performance Engineering    38701-8047   
16  Progressive Shippers Cooperative Association No    53546-9427   
17                             Star Enterprises Inc    99301-3882   
18                   Ryder Truck Rental and Leasing    06457-3834   
19              Buddy Storbeck's Diesel Service Inc    06457-3834   
20                                    Classic Chevy    85257-3102   
21                            Enterprise Rent A Car    60504-7114   
22                                    Race Car Help    78758-7841   
23                                      Capitol KIA    38701-8047   
24                                 C & M Motors Inc    60504-7114   
25                                   Nebo Chevrolet    06457-3834   
26                           Chrysler of Tri-Cities    99301-3882   
27                                    Suburban Ford    53546-9427   

    Revenue Oct 22–Mar 23  Revenue Apr 23–Sep 23  Absolute growth  \
0               3612032.0              5090489.0        1478457.0   
1               3911043.0              5499057.0        1588014.0   
2               8077886.0             11012219.0        2934333.0   
3               8152043.0             11035627.0        2883584.0   
4               4102674.0              5545043.0        1442369.0   
5               3778121.0              5098450.0        1320329.0   
6               7949032.0             10089317.0        2140285.0   
7               4158891.0              5187959.0        1029068.0   
8               3984545.0              4873757.0         889212.0   
9               8476718.0             10347872.0        1871154.0   
10              4351471.0              5276033.0         924562.0   
11              4225424.0              5094952.0         869528.0   
12              4461650.0              5224190.0         762540.0   
13              4241245.0              4962878.0         721633.0   
14              4077222.0              4766085.0         688863.0   
15              8394418.0              9728467.0        1334049.0   
16              9153103.0             10559933.0        1406830.0   
17              8375387.0              9639247.0        1263860.0   
18              8637064.0              9596455.0         959391.0   
19              4036094.0              4459705.0         423611.0   
20              4537847.0              4992161.0         454314.0   
21              4400830.0              4800000.0         399170.0   
22              8879888.0              9627865.0         747977.0   
23              4491596.0              4849059.0         357463.0   
24              4595472.0              4931365.0         335893.0   
25              4536745.0              4749055.0         212310.0   
26              4505493.0              4465360.0         -40133.0   
27              9078015.0              8950083.0        -127932.0   

    Percentage growth  
0           

Median growth rate is 16.9%. The slowest growing was Suburban Ford at -1.41%. Only this deler and Chrysler of Tri-Cities had negative growth.


### 14. Growth rate of companies from the first 6 months to the last 6 months.

In [ ]:
read_sql("""
WITH company_monthly AS (
    SELECT
        company,
        DATE_TRUNC('month', date_of_sale::date) AS month,
        SUM(price) AS revenue
    FROM car_sales
    WHERE date_of_sale::date BETWEEN DATE '2022-10-01' AND DATE '2023-09-30'
    GROUP BY company, DATE_TRUNC('month', date_of_sale::date)
),
company_halfyear AS (
    SELECT
        company,
        CASE
            WHEN month BETWEEN DATE '2022-10-01' AND DATE '2023-03-31'
                THEN 'H1'
            ELSE 'H2'
        END AS period,
        SUM(revenue) AS revenue
    FROM company_monthly
    GROUP BY company,
             CASE
                 WHEN month BETWEEN DATE '2022-10-01' AND DATE '2023-03-31'
                     THEN 'H1'
                 ELSE 'H2'
             END
),
company_pivot AS (
    SELECT
        company,
        SUM(CASE WHEN period = 'H1' THEN revenue ELSE 0 END) AS rev_h1,
        SUM(CASE WHEN period = 'H2' THEN revenue ELSE 0 END) AS rev_h2
    FROM company_halfyear
    GROUP BY company
)
SELECT
    company AS "Company",
    rev_h1  AS "Revenue Oct 22–Mar 23",
    rev_h2  AS "Revenue Apr 23–Sep 23",
    rev_h2 - rev_h1 AS "Absolute growth",
    CASE
        WHEN rev_h1 = 0 THEN NULL
        ELSE (rev_h2 - rev_h1)::numeric / rev_h1
    END AS "Percentage growth"
FROM company_pivot
ORDER BY "Percentage growth" DESC NULLS LAST;

""")

Company  Revenue Oct 22–Mar 23  Revenue Apr 23–Sep 23  Absolute growth  \
0     Plymouth              2950153.0              5216064.0        2265911.0   
1         Saab              1596264.0              2344574.0         748310.0   
2      Porsche              1729171.0              2413684.0         684513.0   
3      Pontiac              4796455.0              6671523.0        1875068.0   
4       Jaguar              1048362.0              1437000.0         388638.0   
5      Lincoln              3281210.0              4414694.0        1133484.0   
6        Honda              4228911.0              5610271.0        1381360.0   
7          BMW              4364904.0              5651524.0        1286620.0   
8     Chrysler              6976037.0              8613739.0        1637702.0   
9        Dodge             10959670.0             13524814.0        2565144.0   
10  Oldsmobile              8320751.0             10207093.0        1886342.0   
11   Chevrolet             11554759.0             14135160.0        2580401.0   
12       Volvo              5036839.0              6131725.0        1094886.0   
13  Volkswagen              8246165.0             10023719.0        1777554.0   
14  Mercedes-B              8040133.0              9570906.0        1530773.0   
15     Mercury              5628400.0              6603313.0         974913.0   
16        Audi              2575979.0              2994445.0         418466.0   
17      Toyota              8071179.0              9245412.0        1174233.0   
18       Lexus              6910073.0              7849631.0         939558.0   
19       Acura              4060252.0              4596297.0         536045.0   
20      Saturn              4338765.0              4847549.0         508784.0   
21      Subaru              2815739.0              3127129.0         311390.0   
22    Cadillac              7091058.0              7760470.0         669412.0   
23  Mitsubishi              7853404.0              8383293.0         529889.0   
24        Jeep              2098154.0              2222190.0         124036.0   
25      Nissan              6793671.0              7192006.0         398335.0   
26    Infiniti              1503811.0              1587311.0          83500.0   
27        Ford             12398817.0             12735815.0         336998.0   
28     Hyundai              1296523.0              1258019.0         -38504.0   
29       Buick              4616340.0              4083313.0        -533027.0   

    Percentage growth  
0            0.768066  
1            0.468788  
2            0.395862  
3            0.390928  
4            0.370710  
5            0.345447  
6            0.326647  
7            0.294765  
8            0.234761  
9            0.234053  
10           0.226703  
11           0.223319  
12           0.217376  
13           0.215561  
14           0.190392  
15           0.173213  
16           0.162449  
17           0.145485  
18           0.135969  
19           0.132023  
20           0.117265  
21           0.110589  
22           0.094402  
23           0.067473  
24           0.059117  
25           0.058633  
26           0.055526  
27           0.027180  
28          -0.029698  
29          -0.115465

Plymouth, in particular, grew by 76.8% from 2022 to 2023, which is much higher than the growth by the second highest-growth company Saab (46.9%).\
Buick's revenue fell by 11.5%, which is quite significant, and Hyundai's revenue fell by 3.0%.\
THis could be a reflection of 

### 15. Dealer market shares in each region

In [ ]:
read_sql("""
WITH region_dealer_counts AS (
    SELECT
        dealer_region,
        dealer_name,
        dealer_no,
        COUNT(*) AS units_sold
    FROM car_sales
    GROUP BY dealer_region, dealer_name, dealer_no
),
region_totals AS (
    SELECT
        dealer_region,
        SUM(units_sold) AS total_units
    FROM region_dealer_counts
    GROUP BY dealer_region
),
region_shares AS (
    SELECT
        rdc.dealer_region,
        rdc.dealer_name,
        rdc.dealer_no,
        rdc.units_sold,
        rdc.units_sold::numeric / rt.total_units AS market_share
    FROM region_dealer_counts rdc
    JOIN region_totals rt
      ON rdc.dealer_region = rt.dealer_region
)
SELECT
    dealer_region AS "Dealer region",
    SUM(market_share * market_share) AS "Herfindahl Index"
FROM region_shares
GROUP BY dealer_region
ORDER BY "Herfindahl Index" DESC;
""")

Dealer region  Herfindahl Index
0    Scottsdale          0.185475
1    Janesville          0.184528
2    Middletown          0.183551
3         Pasco          0.183490
4    Greenville          0.183456
5        Aurora          0.183340
6        Austin          0.155156

Only Austin has a very little competition (0.155), whereas the HI for the rest are very similar and between 0.183 and 0.185.


### 16. Change in composition of engine and transmission in cars solds from 2022 to 2023.

In [ ]:
read_sql("""
WITH engine_trans_year AS (
    SELECT
        EXTRACT(YEAR FROM date_of_sale::date)::int AS year,
        engine,
        transmission,
        COUNT(*) AS units
    FROM car_sales
    GROUP BY
        EXTRACT(YEAR FROM date_of_sale::date)::int,
        engine,
        transmission
),
year_totals AS (
    SELECT
        year,
        SUM(units) AS total_units
    FROM engine_trans_year
    GROUP BY year
)
SELECT
    e.year,
    e.engine,
    e.transmission,
    e.units,
    e.units::numeric / yt.total_units AS share_of_year
FROM engine_trans_year e
JOIN year_totals yt
  ON e.year = yt.year
ORDER BY e.year, e.engine, e.transmission;
""")

year                     engine transmission  units  share_of_year
0  2022  Double Overhead Camshaft         Auto   5653       0.531047
1  2022          Overhead Camshaft       Manual   4992       0.468953
2  2023  Double Overhead Camshaft         Auto   6918       0.521680
3  2023          Overhead Camshaft       Manual   6343       0.478320

Share of Auto and Manual cars did not change much. The market share only shifted by 1% from Auto to Manual.